### PR TITLE
build: handle missing tty gracefully

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

This should hopefully address the recent netflify build update.

### Testing Strategy

You're looking at it.

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
still no luck with netlify it seems

**Comment by hudson-ayers** [comment]
I think the tty issue is a red herring on what is actually wrong. The tty errors were printed even back when the build passed (see https://app.netlify.com/sites/docs-tockosorg/deploys/5f9ad29ca1bd4c0007bed3fd for example).

The netlify build stopped passing on master on the commit where #2175 was merged (we missed this because we thought it was just because the qemu site was down and merged manually, but now I am not so sure).

Another PR that seems to have caused some issues is #2040 , which updated the QEMU submodule. I say this because the build log shows failures for that PR even when other PRs were passing.

Also, interestingly, clearing the netlify cache partially fixes a single build (it gets further along before failing), but you have to reclear the cache before every new commit to see the benefit. So there seem to be two significant issues -- one related to the cache, and the other possibly related to the PR that updated Rust somehow causing builds to timeout even when the cache issue is temporarily fixed.

**Comment by hudson-ayers** [comment]
okay, now I'm getting close! Updating Rust means a new version of rustdoc, and the new version of rustdoc is better at generating warnings -- it found 2 warnings which did not previously exist. Unfortunately, those terrifying bash lines meant to detect warnings and error instead seem to detect warnings and hang when run on the netlify hosted machines. Why? No idea.

But with those warnings fixed, netlify now passes, so long as you manually rebuild the cache. In fact, now that it passes, I am not sure if rebuilding the cache is needed for subsequent builds

**Comment by alistair23** [comment]
WooHoo!

File: boards/components/src/hmac.rs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 //!
 //! Usage
 //! -----
-//! ```rust
+//! ```ignore
```
Comment (alistair23) [inline boards/components/src/hmac.rs:5]: I think it should be `rust,ignore`

Comment (hudson-ayers) [reply boards/components/src/hmac.rs]: it should be just rust, that was a leftover from earlier debugging

**Comment by hudson-ayers** [comment]
For posterity:

this problem has probably existed for awhile, but it only manifests when there are warnings when running `rustdoc`, so we got by not encountering it. 

A minimal reproducible case I could find was this:
```bash
cargo doc 2>&1 | tee /dev/stderr | grep --quiet warning
```
which hangs forever when the output of `cargo doc` includes a warning.

Interestingly, both
```bash
cargo doc 2>&1 | grep --quiet warning
```
and
```bash
cargo doc 2>&1 | tee /dev/stderr | grep warning
```
work correctly in the presence of warnings.

Notably, per `man grep`:
```
-q, --quiet, --silent
      Quiet; do not write anything to standard output.  Exit immediately with zero status if any match is found, even if an error was  detected. 
```

The issue is clearly related to the fact that `grep -q` exits early, closing a pipe when preceding commands are still running.

By monitoring the output of `ps -e` before and after the faulty command hangs, I can see that both the `tee` and `grep` processes do terminate. `cargo doc`, however, does not. So this seems to be some sort of issue with `cargo doc` not correctly handling when its `stdout` or `stderr` is closed. But I fail to understand why this issue only manifests when `tee` is being used between `cargo doc` and the failing command.

My solution was as follows:

1. tee to `/dev/stderr` instead of `/dev/tty`, as there is no `/dev/tty` on the netlify CI machines, so we can't actually see the warnings produced by `cargo doc`  EDIT: It seems this still doesn't work, there is a permissions error. AFAICT this is a docker issue.

2. Rather than use `grep -q`, which exits early, use `grep >> /dev/null`, to avoid issues with `cargo doc` incorrectly handling closed outputs.

3. Fix the `cargo doc` warnings

**Review by hudson-ayers — APPROVED** [review]
I think this is ready for merge

**Review by ppannuto — COMMENTED** [review]
You are a hero for tracking this down. Ironically, I can't approve this because I'm the original PR author, but it looks 👍 to me.

**Review by alistair23 — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Pull request merged**